pyresample/geometry.py

```diff
@@ -320,57 +336,186 @@ def get_bbox_lonlats(self, vertices_per_side: Optional[int] = None, force_clockw
+    def _compute_boundary_mask(self):
+        """Compute valid boundary mask for AreaDefinition(s) with sides out of the Earth disk."""
+        from pyresample.boundary.utils import find_boundary_contour_indices
+        if self._boundary_contour_idx is None:
+            lons, lats = self.get_lonlats()  # all in memory !
+            self._boundary_contour_idx = find_boundary_contour_indices(lons, lats)
+        return self._boundary_contour_idx
+
+    def _get_geographic_sides(self, vertices_per_side: Optional[int] = None) -> tuple:
+        """Return the geographic boundary sides of the current area.
+
+        Args:
+            vertices_per_side:
+                The number of points to provide for each side.
+                By default (None) the full width and height will be provided.
+                If the area object is an AreaDefinition with any corner out of the Earth disk
+                (i.e. full disc geostationary area, Robinson projection, polar projections, ...)
+                by default only 50 points are selected.
 
-    def _get_bbox_elements(self, coord_fun, vertices_per_side: Optional[int] = None) -> tuple:
-        s1_slice, s2_slice, s3_slice, s4_slice = self._get_bbox_slices(vertices_per_side)
-        s1_dim1, s1_dim2 = coord_fun(data_slice=s1_slice)
-        s2_dim1, s2_dim2 = coord_fun(data_slice=s2_slice)
-        s3_dim1, s3_dim2 = coord_fun(data_slice=s3_slice)
-        s4_dim1, s4_dim2 = coord_fun(data_slice=s4_slice)
-        dim1, dim2 = zip(*[(s1_dim1.squeeze(), s1_dim2.squeeze()),
-                           (s2_dim1.squeeze(), s2_dim2.squeeze()),
-                           (s3_dim1.squeeze(), s3_dim2.squeeze()),
-                           (s4_dim1.squeeze(), s4_dim2.squeeze())])
-        if hasattr(dim1[0], 'compute') and da is not None:
-            dim1, dim2 = da.compute(dim1, dim2)
-        clean_dim1, clean_dim2 = self._filter_bbox_nans(dim1, dim2)
-        return clean_dim1, clean_dim2
-
-    def _filter_bbox_nans(
+        Returns:
+            The output structure is a tuple of two lists of four elements each.
+            The first list contains the longitude coordinates.
+            The second list contains the latitude coordinates.
+            Each list element is a numpy array representing a specific side of the geometry.
+            The order of the sides are [top", "right", "bottom", "left"]
+        """
+        is_swath = self.__class__.__name__ == "SwathDefinition"
+        if not is_swath and _is_any_corner_out_of_earth_disk(self):
+            # Geostationary
+            if self.is_geostationary:
+                return self._get_geostationary_boundary_sides(vertices_per_side=vertices_per_side,
+                                                              coordinates="geographic")
+            # Polar Projections, Global Planar Projections (Mollweide, Robinson)
+            # - Retrieve dummy right and left sides
+            if config.get("force_boundary_computations", False):
```

Comment:
This config regulates the behaviour when an AreaDefinition has boundaries on out-of-Earth regions. 

`force_boundary_computations == False` keep the old behaviour of not searching for the actual internal boundary cooordinates, but just select the coordinates at the image border.
For global projections, (i.e. Mollweide, Robinson, ...), polar projections, ... :
- if all image sides are out of Earth, the sides would only have `Inf` values and a `ValueError` is raised further downstream if all values are `Inf`. 
- If some of the image sides are out of Earth, some sides values would be `Inf`, such `Inf` value are discarded further downstream in the processing. If not all values are Inf, a "partial boundary" is returned, however it does not represent the true Earth boundary of the area. 

With `force_boundary_computations == True`, we actually search the actual internal coordinates of the boundary.
To do so, we need to compute in memory one the coordinates array , from here the name `force_*_computations`. 

See my other comment answering to the reason of the try-except logic to understand the rationale of this choice.

Reply:
Maybe we'll keep this discussion on this thread instead of the 3 other comments I made related to the same topic.

Is another way of seeing this as "produce the best possible boundary possible or do the quick/shortcut method"? Where the quick/shortcut is the old way, right? Sorry if I'm getting confused, but this only applies to generating the spherical/lonlat boundaries right? The x/y for AreaDefinitions should always be valid in terms of the x/y coordinates, right? Should we include swaths in this current discussion or wait to confuse me later?

Can we use information from the PROJ/pyproj CRS class (like axis bounds) if they exist to do an even faster method of this where maybe we clip the lon/lats to the defined limits of the CRS? They don't always exist...hhmm but the result is always Inf isn't it? Not `-Inf` and `Inf`...kind of hard to clip then. I could also see us hardcoding some specializations of this for specific projections like we do for geostationary. That would make things a little smarter and faster wouldn't it?

Perhaps a better name for this would be something related to doing the "extra" work to get the best answer rather than phrasing it as "forcing" something. If we can speed up the common case maybe this could default to "on"/True?

Reply:
> Should we include swaths in this current discussion

`SwathDefinition` are not impacted by `force_boundary_computations` flag.

`force_boundary_computations` only impacts AreaDefinition where Earth coordinates only lies inside an internal region of the coordinates array (i.e. geostationary ... but for geostationary we have a smart ad-hoc trick to retrieve the actual internal Earth boundary coordinates without loading the coordinates array into memory).

I looked deeply into PROJ/pyproj CRS and I didn't find another way to determine the actual valid internal coordinate boundary without the trick I implemented in the PR. 
Theoretically one could design ad-hoc methods for each CRS to infer the boundary as we did for geostationary (angle + satellite height) ... but it's a huge, tedious, work.

> This only applies only applies to generating the spherical/lonlat boundaries right?
- `force_boundary_computations=True` is required to perform correct spherical polygon operations 
- `force_boundary_computations=True` will likely fix issues in `get_area_slices` (because of previous bad polygon intersections)
- `force_boundary_computations=True` speeds up some pykdtree operations when `reduce_data=True`. Currently resampling geostationary full disc (i.e. with nearest) using `reduce_data=True` does not reduce the data (because the boundary returned is all `Inf` ...) 

> The x/y for AreaDefinitions should always be valid in terms of the x/y coordinates, right ? 

- `force_boundary_computations=True` is also used in the `area.projection_boundary` method and impacts the gradient resampling, where data reduction is (assumed to be done) in planar projection coordinates. `area.projection_boundary` return the x/y coordinates corresponding to Earth valid coordinates, and so enable the transform of i.e. the actual internal GEO boundary to another CRS --> `reduce_data=True` with gradient was failing for GEO FD and this PR we might have solved the issue (not tested ... I didn't want to also refactor gradient logic :P)

> If we can speed up the common case maybe this could default to "on"/True

I think that the `boundary` and `projection_boundary` methods should return the correct boundaries. 
This will also benefit `get_area_slices` in some cases (remapping to polar projections was causing issues ...)

Then the decision of what should be done in `pykdtree` and `gradient` it's a choice to be based on the following points: 
- Is it fine for some special `AreaDefinition` projections to load the entire coordinates in memory? We go a bit away from the idea of only load the required data with dask ... but on the other side in such special cases ... we were later on loading all coordinates anyway to discover which was valid and which not ... 
- In pykdtree, we will just gain in performance and I don't expect any surprise
- In gradient, there is somewhat lot of assumptions in how data reduction is done and it might be there will be some surprise with some projections, but on the other hand we move forward to fix several issues ...